storage: add design document for additional storage metrics (#24318)
MaterializeInc/materialize: doc/developer/design/20240108_source_metrics_2.md

- _Generally_, how fast we are reading data from upstream?
- Have we committed the source's snapshot?

but fail to answer **two** critical questions users regularly ask about their sources:

1. How much of its snapshot has this source processed?
2. Is this source keeping up with _upstream_? (This is usually related to: can I resize this source?)

This design document proposes metrics that answer these questions, and goes into detail about their
implementation.

## Success Criteria

This design has succeeded when we are able to present these metrics to users in the console,
and they can use them to answer questions about their source's snapshotting and sizing.

## Out of Scope

- Expose new metrics for sinks.
- Solve the _hydration backpressure problem_ (more on this later)
- Expose detailed, _granular_ metrics that are 100% accurate.

## Solution Proposal

### Frontier Aggregation

Sources define progress as frontiers within a source-defined timestamp type. These times
can be partially ordered, and various shapes. In the context of metrics, we want to
be able to _calculate rates_ and _compare progress against a total_ in a way consistent

> across all source.

Reviewer: ```suggestion
across all sources.
```

This means sources will need to implement a way to aggregate their notions of progress into
a single 64-bit unsigned integer. Clients processing this data can then consistently calculate rates
and percentages from these numbers. Note that the _unit_ of the single integer does not need to be
consistent across sources.

#### Kafka frontier aggregation

For example, consider the following kafka source frontiers:
```
# Partition Id: Offset
f1: {0: 100, 1: 150}
```

> A kafka source with this frontier can be said to have processed 240 offets.

Reviewer: This section is a bit misleading as the frontiers of the kafka source do not look like this

#### Postgres frontier aggregation

> The Postgres source has 2 notions of progress. During snapshotting it can sum the number

Reviewer: This also is unclear as to what it means. The postgres source has only one notion of progress, a frontier of LSN

of rows it has processed across all tables. For example:
```
{table1: 110 rows, table2: 15 rows} = 125 rows processed
```

During replication, the source can report the `lsn` as the number of bytes processed.

#### MySQL frontier aggregation

Similar to Postgres, The MySQL source has a rows-per-table defined during snapshotting.

However, during replication, MySQL's source timestamp is partially ordered, instead of just a single
lsn number:
```
# source id: transaction id
{0: 1000th txid, 1: 10th txid} = 1010 transactions processed.

> ```

Reviewer: Same here, the frontier of the MySQL source looks different than this so it could be misleading

Author: tru, ill adjust this section even more

- We will present the metrics as the _percentage of the snapshot Materialize has read_.
- The console will present the user will the memory usage of the replica doing snapshotting, which makes it clear when this scenario is occurring
(This is why memory is included on the above charts).
- In general, once we have read the entire snapshot from upstream, we process the remaining data quite quickly. This is particularly true for non-UPSERT sources.
- Categorically fixing this means implementing _Hydration Backpressure_, which effectively locks Materialize's _read-rate_ to its _process-rate_.
This is considered [out of scope](#out-of-scope) for this design, but, _critically_: if and when _Hydration backpressure_ is implemented, these proposed
metrics _become even more useful_, as opposed to becoming irrelevant.

## Steady-state

The second set of metrics this design document proposes describe the _rate_ at which upstream and Materialize are processing data.
These metrics are designed to answer #2 in [the problem statement](#the-problem).

```
| `outstanding_total`    | [`uint8`] | The total number of outstanding values in the upstream source. The unit is source-specific.       |

> | `processed_total`      | [`uint8`] | The total number of upstream values Materialized has processed. Same unit as `outstanding_total`. |

Reviewer: How will we maintain the ``outstanding_total` number across restarts? I wonder if it makes sense to just have workers report the currently oustanding work as a gauge without a total. What do you think?

Author: It will be a gauge! You are right that `_total` is misleading, as its not a counter; ill adjust the name

Author: changed to `_values`

is a way to capture _how different_ two frontiers are in, in a way that is consistent between source types.

Source frontier differences are used in some form in the metrics proposed in [Snapshot Progress](#snapshot-progress) and
[Steady-state](#steady-state).

#### Kafka frontier differences

For example, consider the following 2 kafka source frontiers:
```
# Partition Id: Offset
f1: {0: 100, 1: 150}
f2: {0: 150, 1: 170}
```

In this case, the source frontier difference is `f2 - f1 = 70 offsets`. Similarly:

```
{0: 100, 1: 100} - {0: 0} = 200 offsets
```

#### Postgres frontier differences

Postgres has two separate definitions:

For the number of rows in each table:
```
{table1: 110 rows, table2: 15 rows} - {table1: 100 rows, table2: 10 rows} = 15 rows
```

And for lsn's during replication:
```
100100 lsn - 100050 lsn = 50 bytes
```

#### MySQL frontier differences

Similar to Postgres, mysql has a rows-per-table difference, as well as a replication difference.

However, during replication, MySQL's source timestamp is partially ordered, instead of just a single
lsn number:

```

> # source id: gtid

Reviewer 2: slight nit: the whole `source_id: transaction_id` format is a GTID

Author: ah i see

f1: {0: 1000th gtid, 1: 0th gtid}
f2: {0: 1010th gtid, 1: 10th gtid}
```
In this case, `f2 - f1 = 20 transactions`

#### Comparisons

As you can see, source frontier differences depend on the source, and are primarily useful for _calculating rates_ and when
_comparing values_. For example:

> - If a source's frontier has moved from `{0: 10}` to `{0: 70}` in 1 minute, Materialize can be said to have processed 1 offset/minute.

Reviewer 3: ```suggestion
- If a source's frontier has moved from `{0: 10}` to `{0: 70}` in 1 minute, Materialize can be said to have processed 1 offset/second.
```
?

Author: thanks!

- If a source's frontier snapshot frontier is `{table1: 100 rows}` and its current frontier is `{table1: 50}`, Materialize can be said
to have processed about 50% of the snapshot.

Additionally, differences are designed to be _lower-bounded estimates_. For example, `{0: 100} - {0: 0, 1: 50}` is 100, as we
choose the lower bound where we assume no offsets past 50 for partition 1 exist.

## Snapshot Progress

The first set of metrics this design document proposes involve _snapshot progress_, which is a _lower bound_ on the _percentage_ of the source's snapshot Materialize
has _read_. These metrics are designed to answer #1 in [the problem statement](#the-problem).

We will introduce 2 new columns in `mz_source_statistics_per_worker`:

```
| `snapshot_total`     | [`uint8`] | The total number of upstream values that are part of the snapshot. |

> | `snapshot_progress   | [`uint8`] | The number of upstream values Materialize has read so far.         |

Reviewer 2: ```suggestion
| `snapshot_progress`   | [`uint8`] | The number of upstream values Materialize has read so far.         |
```

```

The unit of _values_ depends on the source type, and will be _rows_ for MySQL and Postgres, and _offsets_ for kafka.

> These values can be summed across workers and compared (`snapshot_progress / snapshot_progress - snapshot_total`) to produce

Reviewer 3: Is this formula correct? I don't understand it.

Author: you are right, its should be simpler!

a _lower-bound_ estimate on the progress we have made reading the source's snapshot.

### Source frontier difference

> Note that this percentage is calculated using _source frontier difference_, between the `snapshot_total` and `snapshot_progress`.

Reviewer 2: This is a bit misleading since for MySQL & Postgres the frontiers are in terms of LSNs / transactions but `snapshot_total` and `snapshot_progress` will be in terms of row-counts

Author: I tried to capture that there are 2 types of frontiers for pg and mysql in the above sections, is there more context I should add?

Author: Actually, I see your point, ill adjust the structure a bit

For consistency within schema of the `mz_source_statistics_per_worker` table, each worker in the source pipeline is required to
pre-aggregate these values from the possibly complex frontier that represents the total and progress of a snapshot.
This primarily means summing offsets/rows across partitions/tables.

### Source specifics

The `SourceReader` implementation for each source will be required to produce a `snapshot_total`, as well as

> a continuously updated `snapshot_progress` frontier on each worker.

Reviewer: It feels like thinking about source frontier differences was a helpful stepping stone to arrive to the realization that we can have each source define its own way of describing its snapshot progress but beyond that they don't offer any advantages implementation wise.

In particular, if the `SourceRender` trait is extended with the requirement that each source provides a `snapshot_total` and a `snapshot_progress` estimate then we don't gain anything from making the type of these two values some weird frontier that will only ever be used by generic code to compute a `u64` difference. For this reason I propose that each source is obliged to present a `u64` typed `snapshot_total` and `snapshot_progress` which it can calculate in whatever way it wants.

In the same vain I propose that each source exposes an `outstanding_total` gauge (again `u64` typed) which represents the number of outstanding "things" upstream. For each source type a thing might be something else, for postgres it might be the number of outstanding LSN bytes and for kafka the number of oustanding messages. In any case though this number going up represents trouble and is a signal to scale up the cluster.

Author: I like this! will adjust the doc accordingly

#### Kafka

`snapshot_total` can be be trivially exposed by exposing the snapshot frontier already tracked within its source operator,
and summing across partitions. Similarly, `snapshot_progress` can be derived from the operator's output frontier.

#### Postgres and MySQL

`snapshot_total` will need to be calculated, in the unit of rows by performing `SELECT count(*)` on the tables that participate in the snapshot.
Both the Postgres and MySQL implementations will be required to perform this query, per-table, during snapshotting. For RDBMSes, this

> should be cheap.

Reviewer 4: Wait, `SELECT count(*)` is generally decidedly not cheap for RDBMSes: https://wiki.postgresql.org/wiki/Slow_Counting

Perhaps it's fine because we're going to `SELECT *` anyway, but it's definitely not free, and we should think twice about the implications of doing a count(*) first. It might add meaningful overhead to the snapshot time.

Author: (@rjobanp you were also interested in this)

@benesch, I did some research and testing, and while mysql and postgres both have count estimates, they both are problematic; Mysql's can be (as the documentation suggests) 40-50% wrong in innodb, and postgres returns -1 for tables that haven't been `ANALYZED` or vacuumed recently :(

I think we should go ahead and start with `count(*)` over an estimate, but critically: run this count concurrently with the snapshot being processed (I have a local branch that is almost ready that does this, but also offers LD configuration for all these options). This will allow us to at least report the _number of rows we have processed_ before we get a full count, which can still be useful for the user. Ill update the doc to reflect this.

The local branch also lets see in production what real world `count(*)`'s really cost!

Reviewer 4: That seems reasonable! In PostgreSQL, I think you might be able to use `pg_export_snapshot()` and `SET TRANSACTION SNAPSHOT` to allow the second connection you create to count the tables to get the exact same number of rows that the snapshot connection will eventually read.

Author: Yes, thats what I intend to do!

`snapshot_progress` will need to be exposed by the source operators themselves, as the data itself during snapshotting all has
the same frontier. This means the operators will need to track and periodically report a frontier describing the progress they

> have made reading the snapshot. In practice, this number of rows they have read, per-table.

Reviewer 2: ```suggestion
have made reading the snapshot. In practice, this is the number of rows they have read, per-table.
```

These metrics are designed to answer #2 in [the problem statement](#the-problem).

```
| `upstream_rate`      | [`uint8`] | The rate at which data is appearing in the upstream service.       |
| `process_rate`       | [`uint8`] | The rate at which Materialize is processing upstream data.         |
```

This rate has units that depend on the source type (replication bytes for Postgres, offsets for Kafka, gtid's for MySQL) and are
designed to be _compared_. The following charts show what these metrics will look like, depending on whether Materialize is keeping up
or falling behind.

![steady-state](./static/source_metrics_2/steady_state.png)

Note that when our _process rate_ is consistently below the `upstream_rate`, the user knows their source isn't keeping up with upstream.

### Implementation

These rates are _source frontier differences_. Each source's `SourceReader` implementation will
be required to expose a continuous stream of _upstream source frontiers_. This means:

- The Kafka source will periodically fetch metadata and expose per-partition offset high-watermarks.

> - The MySQL source will periodically expose the result of `SELECT @gtid_executed`, which is the gtid of the latest transaction.

Reviewer 2: ```suggestion
- The MySQL source will periodically expose the result of `SELECT @gtid_executed`, which is the set of gtids representing the latest transactions committed.
```

- The Postgres source will periodically expose the result of `SELECT pg_current_wal_lsn()`, which is the lsn of the latest transaction.

Additionally, the source pipeline will periodically invert the latest frontier we have committed for the source, from the Materialize timestamp

> domain to the source-timestamp domain. For MySQL and Postgres sources, this frontier will the `meet` of the subsource frontiers.

Reviewer 5: I don't follow this, can you say more about why we want to do this and how it works?

`upstream_rate` will be derived from the two latest _upstream source frontiers_, by calculating _source frontier difference_ between them.
Similarly `process_rate` will be derived from the two latest committed frontiers.

### Existing metrics

Existing metrics such as `messages_received` and `bytes_received` continue to be useful for understanding the general performance of a given source.
If users find some of them confusing or not useful, we can remove them in the future.

## Minimal Viable Prototype

> N/A for now. This design document is primarily designed to expose

Reviewer 5: Was there more you were going to say here?

Author: oh rip, i reviewed everything else!

## Alternatives

- Expose 100% accurate information about how much of the snapshot we have processed, as opposed to how much we have read.
    - Rejected as it's considered as difficult as implementing Hydration Backpressure.
- Expose progress information in a different format per source type.
    - Rejected as it adds undue complexity in the system catalog, console, and for the user to interpret.
        - In the future, if customers want per-partition information for Kafka sources, we can do so easily using this same general design.
- Consolidate the two metric sets proposed in this design document into one.
    - Rejected due to its complexity.
- Expose more information to users about their sources.
    - This design document is intentionally constrained, to help users answer 2 well-known questions about their sources.

> 

Reviewer 4: I think it's worth calling out here that (I think) https://github.com/MaterializeInc/materialize/issues/23345 is an alternative to (2), in that if we always reclocked to the latest upstream offset, you should be able to observe whether a source is keeping up with upstream by looking for skew between the progress collection and the data collection. The downside of this alternative is that it couples progress to a policy decision about reclocking; what's nice about what you proposed here is that progress tracking is totally independent of reclocking policy.

Author: Its an alternative in the sense that we would get the _upstream_ frontier as part of the pipeline, and we could recover the process frontier by using the reclock collection to invert the mz-timestamp domain data frontier. I don't think that latter part is possible in sql at right now (well, maybe its a non-`STRICT SERIALIZABLE` join between the data and reclock collections? 

Ill definition mention this in the doc

## Open questions

- Are there any edge cases around _source frontier differences_ that can't be reconciled into a relatively meaningful number for the purpose of metrics?
- Are there other scenarios than the ones predicted in the above charts that we aren't capturing?
    - One case might be: people's sources that fall behind upstream just use more memory, while Materialize continues to read as fast as possible.
    The metrics added by this proposal would allow us to easily distinguish between those cases.
- Should `upstream_rate` and `process_rate` be exposed as rates (as described in this design document), or should we instead expose the maximum value, and

> leaving calculating the rate to the client?

Reviewer 4: I like the sound of exposing the maximum value. Seems more in line with the way the other statistics table work? For everything else in the table, the client has to calculate the rate themselves.

Author: Sounds good, ill update this document
